### PR TITLE
Make 'pkg/logs' platform-agnostic

### LINKED ISSUE #6250: Make `pkg/logs` platform-agnostic
/kind user-story

## User Story
As an odo developer, I want to refactor the `pkg/logs` package to make it platform-agnostic, so that we can reuse it for other platforms (e.g., Podman).
At this time, this package is designed for working on Kubernetes only.

## Acceptance Criteria 
- [ ] It should depend on the `platform.Client` interface introduced in #6217
- [ ] It should resolve the actual platform client implementation at dependency injection time
- [ ] It should only work for now with the Kubernetes Client implementation. It should panic if `run-on` is not targeting Kubernetes. #6183 will provide the Podman-specific implementation.

## Links 
- Related Epic (mandatory): https://github.com/redhat-developer/odo/issues/5871

/kind user-story

### PR DESCRIPTION
**Pull request by rm3l**
**What type of PR is this:**
/kind task
/kind code-refactoring

**What does this PR do / why we need it:**
This PR is the outcome of a pair-programming session with @feloy; it makes the `pkg/logs` package agnostic of the platform, 
similar to what was done in #6217.
Similarly, the platform is injected by the dependency injection system, depending on the `--run-on` flag.

**Which issue(s) this PR fixes:**
Fixes #6250

**PR acceptance criteria:**

- [ ] Unit test 

- [ ] Integration test 

- [ ] Documentation 

**How to test changes / Special notes to the reviewer:**
`odo` (and specifically `odo logs`) should work like before :-)

### CODE REVIEW

File: pkg/kclient/interface.go [hunk 1]
```diff
@@ -112,7 +113,7 @@ type ClientInterface interface {
 	GetPodUsingComponentName(componentName string) (*corev1.Pod, error)
 	GetRunningPodFromSelector(selector string) (*corev1.Pod, error)
 	GetPodLogs(podName, containerName string, followLog bool) (io.ReadCloser, error)
-	GetAllPodsInNamespace() (*corev1.PodList, error)
+	GetAllPodsInNamespaceFromSelector(selector string, ns string) (*corev1.PodList, error)
```
Comment (dharmit) [inline pkg/kclient/interface.go:116]: Nit: `s/GetAllPodsInNamespaceFromSelector/GetAllPodsInNamespaceMatchingSelector`

File: pkg/platform/interface.go [hunk 2]
```diff
@@ -1,7 +1,23 @@
...
+	GetPodLogs(podName, containerName string, followLog bool) (io.ReadCloser, error)
+
+	GetPodsMatchingSelector(selector string) (*corev1.PodList, error)
+
+	// GetAllResourcesFromSelector returns all resources of any kind (including CRs) matching the given label selector
+	GetAllResourcesFromSelector(selector string, ns string) ([]unstructured.Unstructured, error)
+
+	GetAllPodsInNamespaceFromSelector(selector string, ns string) (*corev1.PodList, error)
+
+	GetRunningPodFromSelector(selector string) (*corev1.Pod, error)
```
Comment (dharmit) [inline pkg/platform/interface.go:22]: Should the comments be expected here or at the places where the interface is being implemented?

Comment (rm3l) [reply pkg/platform/interface.go]: I've added a few (implementation-independent) comments here as well.

**Comment by rm3l** [comment]
Had to rebase and force-push to benefit from https://github.com/redhat-developer/odo/pull/6260; otherwise tests might not be passing.

File: pkg/platform/interface.go (hunk 2 above)
Comment (dharmit) [reply pkg/platform/interface.go]: That's indeed what we should have in the interface. Implementation specific comment should go into the package/file where it's getting implemented.

Thanks for adding the comments. 👍🏾

**Pull request merged**